velox/substrait/tests/SubstraitVeloxPlanConverterTest.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/Random.h>
+
+#include "velox/exec/tests/utils/OperatorTestBase.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+#include "velox/vector/tests/VectorMaker.h"
+
+#include "velox/substrait/SubstraitToVeloxPlan.h"
+#include "velox/substrait/VeloxToSubstraitPlan.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::test;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::exec::test;
+using namespace facebook::velox::substrait;
+
+class SubstraitVeloxPlanConvertorTest : public OperatorTestBase {
+ protected:
+  /// the function used to make data
+  /// \param size the number of RowVectorPtr.
+  /// \param childSize the size of rowType_, the number of columns
+  /// \param batchSize batch size.  row number
+  /// \return std::vector<RowVectorPtr>
+  std::vector<RowVectorPtr>
+  makeVector(int64_t size, int64_t childSize, int64_t batchSize) {
+    std::vector<RowVectorPtr> vectors;
+    std::mt19937 gen(std::mt19937::default_seed);
+    for (int i = 0; i < size; i++) {
+      std::vector<VectorPtr> children;
+      for (int j = 0; j < childSize; j++) {
+        children.emplace_back(makeFlatVector<int32_t>(batchSize, [&](auto row) {
+          return folly::Random::rand32(INT32_MAX / 4, INT32_MAX / 2, gen);
+        }));
+      }
+
+      std::vector<std::string> childrenNames = rowType_->names();
+      vectors.push_back(makeRowVector({childrenNames}, {children}));
+    }
+    return vectors;
+  };
+
+  void assertVeloxToSubstraitProject(std::vector<RowVectorPtr>&& vectors) {
```

Comment:
yes, you are right, we will plan to add the logic that  convert substrait plan back to velox in the later PR and do the round-trip test to make sure the conversion is loss-less.

Reply:
> we will plan to add the logic that convert substrait plan back to velox in the later PR

Wait, wasn't Substrait-to-Velox conversion added in https://github.com/facebookincubator/velox/pull/1048 ?

Reply:
1048 added partial. will check if can do round-trip test for now, if it can, will add round-trip test now.

Reply:
there need add velox-to-substrait convertor for tablescan node and then can add round-trip test. do you think it's ok added in this PR?

Reply:
Yes, but the ReadRel in substrait used are different in the conversion, in converting from velox to substrait , we use valuesNode->VirtualTable. while in converting from substrait to velox, use local_files->TableScanNode. So I think there need to add the support for tableScan Node in converting from velox to substrait, then we can do  round-trip tests for Filter/Project , your thoughts?

Reply:
Can we add support for converting ValuesNode to Substrait? Does Substrait support something like ValuesNode?

Reply:
yes, we have supported converting valuesNode to Substrait. in substrait, it's ReadRel_VirtualTable

Reply:
while the converting from substrait to valuesNode haven't support. it have supported converting from substrait to TableScanNode. so I think I need to add the support converting from substrait to TableScanNode.

Reply:
Should we add ReadRel_VirtualTable -> ValuesNode conversion? This would allow us to add round trip tests, no?

Reply:
yes, this way also can allow us add round trip tests. Actually there have two ways. one is adding  ReadRel_VirtualTable -> ValuesNode , the other is adding  TableScanNode.->ReadRel_LocalFiles.  Your thought?

Reply:
Ultimately we need both. ReadRel_VirtualTable -> ValuesNode might be simpler, but you would know better.

Reply:
Sure, then I will add one way firstly  for round-trip test and then add the other later.

Reply:
Hi @mbasmanova , I have added the round-trip tests. Please help take a review. Thanks!

Reply:
> @ZJie1 I still see GlobalCommonVarSingleton. I thought we discussed that we don't need it.

We removed the getPlan setPlan in this class, but the setPreFunctionReference and getPreFunctionReference still need to make sure the function id is unique. Your Thoughts?

Reply:
> > @ZJie1 I still see GlobalCommonVarSingleton. I thought we discussed that we don't need it.
> 
> We removed the getPlan setPlan in this class, but the setPreFunctionReference and getPreFunctionReference still need to make sure the function id is unique. Your Thoughts?

I remember that the function Id will be repated when there also have  aggregation Functions  not only scala function. when there only have scala Function, the `function_reference` in substrait plan will be right, while add aggregation, the   `function_reference`  for aggregation function will start from 0 not just add based on the largest scala function id, so we add get/set functions in this class to make the function id works.

Reply:
@ZJie1 I suggest to add setup logic that runs once before any plan conversion. That logic would go over all registered Velox functions, both scalar and aggregate, and assign them unique IDs and whatever else is required by Substrait. The result will be a mapping from function name to ID. The plan converter would take this mapping as an input in the constructor and use it to lookup IDs from function names.

BTW, do we need different IDs for functions with same name, but different inputs? Do you have a reference to Substrait docs which explain what is function ID and how should one construct it?

Reply:
> @ZJie1 I suggest to add setup logic that runs once before any plan conversion. That logic would go over all registered Velox functions, both scalar and aggregate, and assign them unique IDs and whatever else is required by Substrait. The result will be a mapping from function name to ID. The plan converter would take this mapping as an input in the constructor and use it to lookup IDs from function names.

Could we add this logic in follow up PR  and then remove this class in that pr or is this part a small changes that should be added here?

Reply:
> BTW, do we need different IDs for functions with same name, but different inputs? Do you have a reference to Substrait docs which explain what is function ID and how should one construct it?

yes, we need different IDs for this case.  There have an example in substrait java https://github.com/substrait-io/substrait-java

Reply:
I would like to avoid adding GlobalCommonVarSingleton class. It might be OK to hard-code a handful of functions and their IDs if you'd like to work on proper function support in a follow-up PR.

Reply:
Sure, will remove it.